Add placeholder advice (#76)
canonical/praecepta: en/index.md

Reviewer: I like this recommendation in general. But I think you should also add somewhere how placeholder variables should be formatted generally (e.g., in angled brackets, separated by an underscore). I don't think we have a formal standard in our style guide yet for that.

Author: We don't have a formal standard for that and it is something I did consider. The problem is, it depends on context. <xxx> is deadly to anything web related, {xxx} causes problems with some build tools, _xxx_ is too easy to miss and probably causes problems with something. So the problem is, I think the sensible default is different depending on the docs you are writing.

Author: ^ proving my point, the missing term there is `<xxx>`

Reviewer 2: I agree that there is no formal standard and that we shouldn't suggest that there is.

My only query would be whether something like "Placeholder delimiters and terms should be appropriate to the context of the documentation in which they appear" needs to be in this section. That might seem vague but it disabuses the reader of the notion that there is a standard that we have chosen not to disclose or forgotten to disclose. Also: it's true.

Author: Added some words

...
```

### Placeholders in code blocks

There are many times when it will be necessary for the user to supply terms or
data to a command which cannot be known in advance, for example IP addresses or

> names. It is common to substitute such values with a placeholder, where the

Reviewer 2: I think what was meant here is that the *author* of the doc may not know the values (that a given user will enter) in advance, not the user.

The suggested change seems fine to me and avoids having to make some painstaking distinction between what the user knows and what the author knows (although that does sound like fun...).

> names. It is common to substitute such values with a placeholder, where the

Author: Well, it was sort of both, but I agree this is simpler

placeholder consists of delimiters that contain terms that suggest the

> necessary input. For example:

Author: @edibotopic

> necessary input. For example:

Reviewer 2: I think that the suggested change is nicer and cleaner.

However, is it still clear to the reader that the placeholder is constituted by both the delimiters and the term(s) inside?

Perhaps just adding "[...] a placeholder, **consisting of terms within delimiters**, where the placeholder [...]" (bold only for emphasis).

I think it's good to say explicitly what the placeholder is somewhere. It's not just words and it's not just delimiters, it's both, and I don't think this is always obvious to people.

I don't think we can say what symbols the delimiters need to be but we can say there are delimiters.

> necessary input. For example:

Author: @edibotopic agree. That version seems familiar somehow :thinking:

> necessary input. For example:

Author: I'm going to confuse everyone by accepting this and then rewriting it.

```
lxc delete <instance_name>/<snapshot_name>
```

> Here, the reader is expected to substitute their values for the placeholders.
> To minimise the risk of errors, instruct the reader that such values need to be
> substituted, especially when the first placeholder is referenced.
> 
> This approach is in common use and is effective for short commands, but is less
> suitable for a long block of code where the user then has to edit a multi-line
> command and search through to find the placeholders to replace. A useful
> approach in these circumstances is to adjust the code to define user-supplied
> data as environmental variables. For example:

Reviewer: I find this to be quite wordy for something that's pretty basic. How about something like:

```suggestion
This example contains a short command, and the placeholders are clearly visible 
to the user. 

However, in longer code blocks, the placeholders become 
more difficult to manage and easier for a user to overlook. For long code 
blocks, consider defining the placeholders as environmental 
variables instead. For example:
```

Author: I don't think we should omit the part about instructing the reader.  Consider the audience here, who are very likely to make incorrect assumptions about what is pretty basic. If we use placeholder then we should make sure we tell people and not expect them to guess.
i will take a stab at rewording this though

````
Set the desired channel:

```

> CHANNEL=1.30/stable

Reviewer: Should your example include a placeholder instead of a real value?
```suggestion
CHANNEL=<channel_name>
```

Author: yes. The idea is that the text would explain what the user needs to do. In this case a sensible real value is there. I prefer this approach because:
- if the reader blindy copies everything it still works
- it is also suitable for automated testing

Author: I mean of course, no. the example is correct. I'm right.

```

Then proceed to fetch the required charms:

```
juju download easyrsa --channel=$CHANNEL
juju download kubernetes-worker --channel=$CHANNEL
...
```

````

This has the following advantages:

 - Separates user-supplied data from commands, making it easier to reference and explain
 - Enables blocks of code to be copied without modification

>  - Reduces the chance of users making mistakes when editing the commands

Reviewer: I'm not sure if this information is needed. The purpose of this guidance feels overly-supported for a style guide. If you think this information is necessary, maybe it could be moved up to before the code block? If it's at the end here, it's easier to miss

Reviewer: Oh, but if you keep it, I think you should at least denote what "This" refers to so it reads less ambiguous.

```suggestion
This approach has the following advantages:

 - Separates user-supplied data from commands, making it easier to reference and explain
 - Enables blocks of code to be copied without modification
 - Reduces the chance of users making mistakes when editing the commands
```

...
```

### Placeholders in code blocks

There are many situations where it's necessary for users to provide their own
information in a code block, such as IP addresses or names. It's common to
substitute these values with a placeholder, consisting of terms within
delimiters, representing the value to be replaced. For example:

```
lxc delete <instance_name>/<snapshot_name>
```

Here, the reader is expected to substitute their values for the placeholders.
To minimise the risk of errors, instruct the reader that such values need to be
substituted, especially when the first placeholder is referenced. There is no
set style for the delimiter: the author should choose something unlikely to be
confusing in the context, and use it consistently.

> However, in longer code blocks the placeholders become more difficult to manage

Author: Fine. It does mean you will have to re-approve though ;)
